### PR TITLE
Port gfp

### PR DESCRIPTION
**Pull request by plabanca**
Ported gfp_generic and gfp_test from https://github.com/ethereum/go-ethereum/tree/master/crypto/bn256/cloudflare

### CODE REVIEW

File: src/gfp.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#ifndef BN256_GFP_H
+#define BN256_GFP_H
+
+#include <cstdint>
+#include <array>
+
+namespace bn256 {
+
+    typedef std::array<uint64_t, 4> uint64_array_4_t;
+    typedef uint64_array_4_t gfp;
```
Comment (heifner) [inline src/gfp.h]: `gfp` allows for aggregate initialization which is nice, but we need to be careful with default initialization as those will be of indeterminate value. I didn't see any issues with this, just a reminder to keep it in mind.

File: src/gfp12.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+namespace bn256 {
+
+    // gfp12 implements the field of size p¹² as a quadratic extension of gfp6
+    // where ω²=τ.
+    struct gfp12 {
+        // value is xω + y
+        gfp6 x;
+        gfp6 y;
+
+        const gfp12& set(gfp12 &a) {
```
Comment (heifner) [inline src/gfp12.h]: Please be consistent with `gfp12& a` instead of `gfp12 &a` throughout the code.

File: src/gfp6.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    struct gfp6 {
+        // value is xτ² + yτ + z
+        gfp2 x;
+        gfp2 y;
+        gfp2 z;
+        
+        void set(gfp6 &a) {
+            x.set(a.x);
+            y.set(a.y);
+            y.set(a.z);
```
Comment (heifner) [inline src/gfp6.h]: `z.set`

**Review by heifner — CHANGES_REQUESTED** [review]

File: src/gfp12.h (hunk 2 above)
Comment (plabanca) [reply src/gfp12.h]: resolved with commit 185b5dc47da805da953d51ecf48b6943d71c9fe1

File: src/gfp6.h (hunk 3 above)
Comment (plabanca) [reply src/gfp6.h]: resolved with commit 185b5dc47da805da953d51ecf48b6943d71c9fe1

File: src/gfp.h (hunk 1 above)
Comment (plabanca) [reply src/gfp.h]: Added default initialization to gfp2 member variables with commit https://github.com/AntelopeIO/bn256/commit/185b5dc47da805da953d51ecf48b6943d71c9fe1

File: src/constants.h [hunk 4]
```diff
@@ -0,0 +1,96 @@
+#ifndef BN256_CONSTANTS_H
```
Comment (linh2931) [inline src/constants.h]: In leap, we use `#pragma once` for including guard. This comment applied to all header files.

File: CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.8)
```
Comment (linh2931) [inline CMakeLists.txt:1]: In leap, top level `CMakeLists.txt `usually includes subdirectories and actual `CmakeLists.txt` files reside in the source directories.

File: src/gfp.h (hunk 1 above)
Comment (linh2931) [inline src/gfp.h]: Define `gfp` as a `struct` so functions in gfp.go can be easily implemented. Why `newGFp`, `Invert`, `montEncode` and `montDecode` are not ported? `Marshal` and `Unmarshal` might need to be ported too.

File: src/gfp12.h [hunk 6]
```diff
@@ -0,0 +1,29 @@
+#ifndef BN256_GFP12_H
```
Comment (linh2931) [inline src/gfp12.h]: What about other functions in gfp2.go which are not ported? Will they be moved over after initial testing?

File: src/gfp12.h [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+namespace bn256 {
+
+    // gfp12 implements the field of size p¹² as a quadratic extension of gfp6
+    // where ω²=τ.
+    struct gfp12 {
+        // value is xω + y
+        gfp6 x;
+        gfp6 y;
+
+        const gfp12& set(gfp12& a) {
```
Comment (linh2931) [inline src/gfp12.h]: Change `gfp12& a` to `const gfp12& a`

File: src/constants.h [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+#define BN256_CONSTANTS_H
+
+#include <gfp.h>
+#include <gfp2.h>
+#include <gfp6.h>
+#include <boost/multiprecision/cpp_int.hpp>
+
+namespace bn256 {
+    // u is the BN parameter.
+    static const auto u = boost::multiprecision::uint128_t("4965661367192848881");
```
Comment (linh2931) [inline src/constants.h]: Can this be a `constexpr`?

File: src/gfp_generic.cpp [hunk 9]
```diff
@@ -0,0 +1,198 @@
+
```
Comment (linh2931) [inline src/gfp_generic.cpp]: Remove empty line.

File: src/gfp_generic.cpp [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+    typedef std::array<uint64_t, 8> uint64_array_8_t;
+    typedef std::array<uint64_t, 18> uint64_array_18_t;
+    typedef std::array<uint64_t, 32> uint64_array_32_t;
+
+    static void gfp_carry(gfp& a, uint64_t head){
+        gfp b{};
+        uint64_t carry = 0;
+        for (const auto& pi : p2) {
+            const auto i = &pi - &p2[0];
```
Comment (linh2931) [inline src/gfp_generic.cpp]: Why not just `for (auto i = 0; i < p2.size(); ++i)` and replace `p2` to `p2[i] `below?

File: src/gfp_test.cpp [hunk 11]
```diff
@@ -0,0 +1,86 @@
+#include <constants.h>
```
Comment (linh2931) [inline src/gfp_test.cpp]: Tests should be eventually integrated into CTEST so that they can be run automatically in CICD.

More corner cases should be added in the future.

**Review by linh2931 — CHANGES_REQUESTED** [review]

File: src/constants.h (hunk 4 above)
Comment (plabanca) [reply src/constants.h]: I changed header include guards with commit https://github.com/AntelopeIO/bn256/pull/1/commits/a5ac4c6997ca11133461deb17f081f5cd318587c

File: CMakeLists.txt (hunk 5 above)
Comment (plabanca) [reply CMakeLists.txt:1]: refactored project per your comment: https://github.com/AntelopeIO/bn256/pull/1/commits/f3870d5ee4aa9a93d08f1e0fb83c870b68ff76c4

File: src/gfp12.h (hunk 7 above)
Comment (plabanca) [reply src/gfp12.h]: This was changed in a commit in another branch, I moved the commit to this branch.

File: src/gfp12.h (hunk 6 above)
Comment (plabanca) [reply src/gfp12.h]: They were implemented in commits in another branch, I moved the commits to this branch.

File: src/gfp.h (hunk 1 above)
Comment (plabanca) [reply src/gfp.h]: resolved with commit: https://github.com/AntelopeIO/bn256/pull/1/commits/8c6ecfb9b0b16fb5190591eb29af13fbcd2e6ebc

File: src/constants.h (hunk 8 above)
Comment (plabanca) [reply src/constants.h]: Yes see commit: https://github.com/AntelopeIO/bn256/pull/1/commits/6704582113e24cea0643ea6e3bc8a37639caa038

File: src/gfp_generic.cpp (hunk 9 above)
Comment (plabanca) [reply src/gfp_generic.cpp]: Done, commit: https://github.com/AntelopeIO/bn256/pull/1/commits/bf15202c2237f7f20985d2b15f012e2ed34e81d2

File: src/gfp_generic.cpp (hunk 10 above)
Comment (plabanca) [reply src/gfp_generic.cpp]: Done. commit: https://github.com/AntelopeIO/bn256/pull/1/commits/58847c2d9049688f99b05380bafc3863c8150894

File: src/gfp_test.cpp (hunk 11 above)
Comment (plabanca) [reply src/gfp_test.cpp]: Done, commit: https://github.com/AntelopeIO/bn256/pull/1/commits/f3870d5ee4aa9a93d08f1e0fb83c870b68ff76c4

**Review by linh2931 — APPROVED** [review]

**Review by heifner — APPROVED** [review]
